lib/ams_lazy_relationships/extensions/reflection.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+# There is a general problem inside AMS related to custom association finder
+# combined with `include_data` setting:
+#
+#   class BlogPost3Serializer < BaseTestSerializer
+#     belongs_to :category do
+#       include_data :if_sideloaded
+#       object.categories.last
+#     end
+#   end
+#
+# The problem is that `belongs_to` block will be fully evaluated each time for
+# each object, and only after that AMS is able to take into account
+# `include_data` mode -
+# https://github.com/rails-api/active_model_serializers/blob/v0.10.10/lib/active_model/serializer/reflection.rb#L162-L163
+#
+#   def value(serializer, include_slice)
+#     # ...
+#     block_value = instance_exec(serializer, &block) if block
+#     return unless include_data?(include_slice)
+#     # ...
+#   end
+#
+# That causing redundant (and so huge potentially!) SQL queries and AR objects
+# allocation when `include_data` appears to be `false` but `belongs_to` block
+# defines instant (not a kind of AR::Relation) custom association finder.
+#
+# Described problem is a very specific use case for pure AMS applications.
+# The bad news is that `ams_lazy_relationships` is always utilize the
+# association block -
+# https://github.com/Bajena/ams_lazy_relationships/blob/v0.2.0/lib/ams_lazy_relationships/core/relationship_wrapper_methods.rb#L32-L36
+#
+# def define_lazy_association(type, name, options, block)
+#   #...
+#   block ||= lambda do |serializer|
+#     serializer.public_send("lazy_#{name}")
+#   end
+#
+#   public_send(type, name.to_sym, real_relationship_options, &block)
+#   #...
+# end
+#
+# This way we break `include_data` optimizations for the host application.
+#
+# In order to overcome that we are forced to monkey-patch
+# `AmsLazyRelationships::Extensions::Reflection#value` method and make it to be
+# ready for Proc returned by association block. This way we will use a kind of
+#
+#   block ||= lambda do |serializer|
+#     -> { serializer.public_send("lazy_#{name}") }
+#   end
+#
+# as association block, then AMS will evaluate it, get the value of `include_data`
+# setting, make a decision do we need to continue with that association, if so -
+# will finally evaluate the proc with lazy relationship inside it.
+
+module AmsLazyRelationships
+  module Extensions
+    module Reflection
+      def value(*)
+        case (block_value = super)
+        when Proc, Method then block_value.call
+        else block_value
+        end
+      end
+    end
+  end
+end
+
+::ActiveModel::Serializer::Reflection.prepend AmsLazyRelationships::Extensions::Reflection
```

Comment:
I'd consider applying this patch only to AMS versions >= 0.10.3. 
Nobody using lower versions of AMS is probably going to return lambdas from their association blocks anyways as they won't benefit from that anyhow.

Reply:
I thought a lot about that from the beginning, but this way we need to apply the same condition in default `define_lazy_association` block.

Even worse, readme description will be more complex as well since for lower AMS versions the proc-like custom finder will instantly fail. I.e. we propose the snippet of code which will crush on AMS version accepted by `gemspec` file..

As an option, we can move all `include_data` stuff in separate lib file. `require 'ams_lazy_relationships/efficient_finders'` will engage monkey-patch and proc-like default finder, without said `require` nothing will be applied.

How you think?

Reply:
> As an option, we can move all `include_data` stuff in separate lib file.

Notice that this way `ams_lazy_relationships` will still silently produce a huge problems to consumers due to cancelled effect from `include_data` optimizations in sense of DB hits and memory usage.

Reply:
Ok, you're right, doing it the way I suggested generates more problems than it solves. Let's keep it then. But in such case let's make sure that all AMS versions are tested for proc evaluation (specs are skipped for lower AMS versions in https://github.com/Bajena/ams_lazy_relationships/pull/47/files#diff-6e72b60a2fb8b930cc9741785114a948R871).

Reply:
They are tested already due to the new default `define_lazy_association` block, but I'll add an  accented test with proc-like custom finder )